wagtail/wagtailadmin/search.py

```diff
@@ -0,0 +1,103 @@
+from __future__ import unicode_literals
+
+from django.forms import MediaDefiningClass, Media
+from django.forms.utils import flatatt
+from django.utils.text import slugify
+from django.utils.safestring import mark_safe
+from django.utils.six import text_type
+
+from django.utils.six import with_metaclass
+
+from wagtail.utils.compat import render_to_string
+from wagtail.wagtailcore import hooks
+from wagtail.wagtailadmin.forms import SearchForm
+
+
+class SearchArea(with_metaclass(MediaDefiningClass)):
+    template = 'wagtailadmin/shared/search_area.html'
+
+    def __init__(self, label, url, name=None, classnames='', attrs=None, order=1000):
+        self.label = label
+        self.url = url
+        self.classnames = classnames
+        self.name = (name or slugify(text_type(label)))
+        self.order = order
+
+        if attrs:
+            self.attr_string = flatatt(attrs)
+        else:
+            self.attr_string = ""
+
+    def is_shown(self, request):
+        """
+        Whether this search area should be shown for the given request; permission
+        checks etc should go here. By default, search areas are shown all the time
+        """
+        return True
+
+    def is_active(self, request):
+        return request.path.startswith(self.url)
+
+    def render_html(self, request, query):
+        return render_to_string(self.template, {
+            'name': self.name,
+            'url': self.url,
+            'classnames': self.classnames,
+            'attr_string': self.attr_string,
+            'label': self.label,
+            'active': self.is_active(request),
+            'query_string': query
+        }, request=request)
+
+
+class Search(object):
+    def __init__(self, register_hook_name, construct_hook_name=None):
+        self.register_hook_name = register_hook_name
+        self.construct_hook_name = construct_hook_name
+        # _registered_search_areas will be populated on first access to the
+        # registered_search_areas property. We can't populate it in __init__ because
+        # we can't rely on all hooks modules to have been imported at the point that
+        # we create the admin_search and settings_search instances
+        self._registered_search_areas = None
+
+    @property
+    def registered_search_areas(self):
+        if self._registered_search_areas is None:
+            self._registered_search_areas = [fn() for fn in hooks.get_hooks(self.register_hook_name)]
+        return self._registered_search_areas
```

Comment:
Also, if the `__lt__()` method were defined on the `SearchArea` class, and the [`total_ordering`](https://docs.python.org/3/library/functools.html#functools.total_ordering) class decorator added, the `SearchArea`s could be ordered without using `key=lambda s: s.order` at all
```
def __lt__(self, other):
    return (self.order, self.label) < (other.order, other.label)
def __eq__(self, other):
    return (self.order, self.label) == (other.order, other.label)
```